specs/phase1/fraud-proofs.md

```diff
@@ -29,42 +32,264 @@ This document describes the shard transition function and fraud proofs as part o
+            beacon_parent_root=beacon_parent_root,
+            signed_block=proposal,
+        )
+
+    return proposal, shard_state, shard_parent_root
+```
 
-* Initialize `proposals = []`, `shard_states = []`, `shard_state = state.shard_states[shard][-1]`, `start_slot = shard_state.slot`.
-* For `slot in get_offset_slots(state, start_slot)`, do the following:
-    * Look for all valid proposals for `slot`; that is, a Bytes `proposal` where `shard_state_transition(shard, slot, shard_state, get_block_root_at_slot(state, state.slot - 1), get_shard_proposer_index(state, shard, slot), proposal)` returns a result and does not throw an exception. Let `choices` be the set of non-empty valid proposals you discover.
-    * If `len(choices) == 0`, do `proposals.append(make_empty_proposal(shard_state, slot))`
-    * If `len(choices) == 1`, do `proposals.append(choices[0])`
-    * If `len(choices) > 1`, let `winning_proposal` be the proposal with the largest number of total attestations from slots in `state.shard_next_slots[shard]....slot-1` supporting it or any of its descendants, breaking ties by choosing the first proposal locally seen. Do `proposals.append(winning_proposal)`.
-    * If `proposals[-1]` is NOT an empty proposal, set `shard_state = shard_state_transition(shard, slot, shard_state, get_block_root_at_slot(state, state.slot - 1), get_shard_proposer_index(state, shard, slot), proposals[-1])` and do `shard_states.append(shard_state)`. If it is an empty proposal, leave `shard_state` unchanged.
+```python
+def get_shard_state_transition_result(
+    beacon_state: BeaconState,
+    shard: Shard,
+    shard_blocks: Sequence[SignedShardBlock],
+    validate_result: bool=True,
+) -> Tuple[Sequence[SignedShardBlock], Sequence[ShardState], Sequence[Root]]:
+    proposals = []
+    shard_states = []
+    shard_state = beacon_state.shard_states[shard].copy()
+    shard_parent_root = beacon_state.shard_states[shard].latest_block_root
+    for slot in get_offset_slots(beacon_state, shard):
+        proposal, shard_state, shard_parent_root = get_proposal_at_slot(
+            beacon_state=beacon_state,
+            shard_state=shard_state,
+            slot=slot,
+            shard=shard,
+            shard_parent_root=shard_parent_root,
+            shard_blocks=shard_blocks,
+            validate_result=validate_result,
+        )
+        shard_states.append(shard_state)
+        proposals.append(proposal)
+
+    shard_data_roots = compute_shard_data_roots(proposals)
+
+    return proposals, shard_states, shard_data_roots
+```
+
+### Make attestations
+
+Suppose you are a committee member on shard `shard` at slot `current_slot` and you have received shard blocks `shard_blocks`. Let `state` be the head beacon state you are building on, and let `QUARTER_PERIOD = SECONDS_PER_SLOT // 4`. `2 * QUARTER_PERIOD` seconds into slot `current_slot`, run `get_shard_transition(beacon_state, shard, shard_blocks)` to get `shard_transition`.
```

Comment:
```suggestion
Suppose you are a committee member on shard `shard` at slot `current_slot` and you have received shard blocks `shard_blocks` since the latest successful crosslink for `shard` into the beacon chain. Let `state` be the head beacon state you are building on, and let `QUARTER_PERIOD = SECONDS_PER_SLOT // 4`. `2 * QUARTER_PERIOD` seconds into slot `current_slot`, run `get_shard_transition(beacon_state, shard, shard_blocks)` to get `shard_transition`.
```